File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/RouteRefresh.java
package com.mapbox.services.android.navigation.v5.navigation;

import android.text.TextUtils;

import com.mapbox.api.directions.v5.models.DirectionsRoute;
import com.mapbox.api.directionsrefresh.v1.MapboxDirectionsRefresh;
import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;

import timber.log.Timber;

/**
 * This class allows the developer to interact with the Directions Refresh API, receiving updated
 * annotations for a route previously requested with the enableRefresh flag.
 */
public final class RouteRefresh {

  private static final String INVALID_DIRECTIONS_ROUTE = "RouteProgress passed has invalid DirectionsRoute";
  private final String accessToken;
  private RefreshCallback refreshCallback;

  /**
   * Creates a {@link RouteRefresh} object which calls the {@link RefreshCallback} with updated
   * routes.
   *
   * @param accessToken     mapbox access token
   * @param refreshCallback to call with updated routes
   * @deprecated use {@link RouteRefresh(String)} instead
   */
  @Deprecated
  public RouteRefresh(String accessToken, RefreshCallback refreshCallback) {
    this.accessToken = accessToken;
    this.refreshCallback = refreshCallback;
  }

  /**
   * Creates a {@link RouteRefresh} object.
   *
   * @param accessToken mapbox access token
   */
  public RouteRefresh(String accessToken) {
    this.accessToken = accessToken;
  }

  /**
   * Refreshes the {@link DirectionsRoute} included in the {@link RouteProgress} and returns it
   * to the callback that was originally passed in. The client will then have to update their
   * {@link DirectionsRoute} with the {@link com.mapbox.api.directions.v5.models.LegAnnotation}s
   * returned in this response. The leg annotations start at the current leg index of the
   * {@link RouteProgress}
   * <p>
   * Note that if {@link RefreshCallback} is not passed in {@link RouteRefresh(String, RefreshCallback)} this call
   * will be ignored.
   * </p>
   *
   * @param routeProgress to refresh via the route and current leg index
   * @deprecated use {@link RouteRefresh#refresh(RouteProgress, RefreshCallback)} instead
   */
  @Deprecated
  public void refresh(RouteProgress routeProgress) {
    if (refreshCallback == null) {
      Timber.e("RefreshCallback cannot be null.");
      return;
    }
    refresh(routeProgress.directionsRoute(), routeProgress.legIndex(), refreshCallback);
  }

  /**
   * Refreshes the {@link DirectionsRoute} included in the {@link RouteProgress} and returns it
   * to the callback that was originally passed in. The client will then have to update their
   * {@link DirectionsRoute} with the {@link com.mapbox.api.directions.v5.models.LegAnnotation}s
   * returned in this response. The leg annotations start at the current leg index of the
   * {@link RouteProgress}
   *
   * @param routeProgress   to refresh via the route and current leg index
   * @param refreshCallback to call with updated routes
   */
  public void refresh(RouteProgress routeProgress, RefreshCallback refreshCallback) {
    refresh(routeProgress.directionsRoute(), routeProgress.legIndex(), refreshCallback);
  }

  private void refresh(final DirectionsRoute directionsRoute, final int legIndex, RefreshCallback refreshCallback) {
    if (isInvalid(directionsRoute, refreshCallback)) {
      return;
    }
    MapboxDirectionsRefresh.builder()
      .requestId(directionsRoute.routeOptions().requestUuid())
      .routeIndex(Integer.valueOf(directionsRoute.routeIndex()))
      .legIndex(legIndex)
      .accessToken(accessToken)
      .build().enqueueCall(new RouteRefreshCallback(directionsRoute, legIndex, refreshCallback));
  }

  private boolean isInvalid(DirectionsRoute directionsRoute, RefreshCallback refreshCallback) {
    String requestUuid = directionsRoute.routeOptions().requestUuid();
    if (TextUtils.isEmpty(requestUuid) || directionsRoute.routeIndex() == null) {
      refreshCallback.onError(new RefreshError(INVALID_DIRECTIONS_ROUTE));
      return true;
    }
    return false;
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/RouteRefresh.kt
package com.mapbox.services.android.navigation.v5.navigation

import android.content.Context
import com.mapbox.api.directions.v5.models.DirectionsRoute
import com.mapbox.api.directionsrefresh.v1.MapboxDirectionsRefresh
import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress
import com.mapbox.services.android.navigation.v5.utils.extensions.ifNonNull

/**
 * This class allows the developer to interact with the Directions Refresh API, receiving updated
 * annotations for a route previously requested with the enableRefresh flag.
 *
 * Creates a [RouteRefresh] object.
 *
 * @param accessToken mapbox access token
 */
class RouteRefresh(private val accessToken: String, private val context: Context) {

    companion object {
        private const val INVALID_DIRECTIONS_ROUTE =
            "RouteProgress passed has invalid DirectionsRoute"
    }

    /**
     * Refreshes the [DirectionsRoute] included in the [RouteProgress] and returns it
     * to the callback that was originally passed in. The client will then have to update their
     * [DirectionsRoute] with the [com.mapbox.api.directions.v5.models.LegAnnotation]s
     * returned in this response. The leg annotations start at the current leg index of the
     * [RouteProgress]
     *
     * @param routeProgress to refresh via the route and current leg index
     * @param refreshCallback to call with updated routes
     */
    fun refresh(routeProgress: RouteProgress, refreshCallback: RefreshCallback) {
        ifNonNull(routeProgress.directionsRoute(), routeProgress.legIndex()) { directionsRoute, legIndex ->
            refresh(directionsRoute, legIndex, refreshCallback)
        }
    }

    private fun refresh(
        directionsRoute: DirectionsRoute,
        legIndex: Int,
        refreshCallback: RefreshCallback
    ) {
        if (isInvalid(directionsRoute, refreshCallback)) {
            return
        }
        val builder: MapboxDirectionsRefresh.Builder = MapboxDirectionsRefresh.builder()
            .requestId(directionsRoute.routeOptions()?.requestUuid())
        directionsRoute.routeIndex()?.let { routeIndex ->
            builder.routeIndex(routeIndex.toInt())
        }
        builder.legIndex(legIndex)
            .accessToken(accessToken)
            .build().enqueueCall(RouteRefreshCallback(directionsRoute, legIndex, refreshCallback))
    }

    private fun isInvalid(
        directionsRoute: DirectionsRoute,
        refreshCallback: RefreshCallback
    ): Boolean {
        val requestUuid = directionsRoute.routeOptions()?.requestUuid() ?: ""
        if (requestUuid.isEmpty() || directionsRoute.routeIndex() == null) {
            refreshCallback.onError(RefreshError(INVALID_DIRECTIONS_ROUTE))
            return true
        }
        return false
    }
}
